pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/pendingack/PendingAckPersistentTest.java

```diff
@@ -100,6 +104,79 @@ protected void cleanup() {
+        // 1. Prepare and make sure the consumer can be built successfully.
+        String topic = NAMESPACE1 + "/testUnloadSubscriptionWhenFailedInitPendingAck";
+        @Cleanup
+        Consumer<byte[]> consumer1 = pulsarClient.newConsumer()
+                .subscriptionName("subName1")
+                .topic(topic)
+                .subscribe();
+        // 2. Mock a transactionPendingAckStoreProvider to test building consumer
+        // failing at transactionPendingAckStoreProvider::checkInitializedBefore.
+        Field transactionPendingAckStoreProviderField = PulsarService.class
+                .getDeclaredField("transactionPendingAckStoreProvider");
+        transactionPendingAckStoreProviderField.setAccessible(true);
+        TransactionPendingAckStoreProvider pendingAckStoreProvider =
+                (TransactionPendingAckStoreProvider) transactionPendingAckStoreProviderField
+                        .get(pulsarServiceList.get(0));
+        TransactionPendingAckStoreProvider mockProvider = mock(pendingAckStoreProvider.getClass());
+        // 3. Test retryable exception when checkInitializedBefore:
+        // The consumer will be built successfully after one time retry.
+        when(mockProvider.checkInitializedBefore(any()))
+                // First, the method checkInitializedBefore will fail with a retryable exception.
+                .thenReturn(FutureUtil.failedFuture(new ManagedLedgerException("mock fail initialize")))
+                // Then, the method will be executed successfully.
+                .thenReturn(CompletableFuture.completedFuture(false));
+        transactionPendingAckStoreProviderField.set(pulsarServiceList.get(0), mockProvider);
+        @Cleanup
+        Consumer<byte[]> consumer2 = pulsarClient.newConsumer()
+                .subscriptionName("subName2")
+                .topic(topic)
+                .subscribe();
+
+        // 4. Test retryable exception when newPendingAckStore:
+        // The consumer will be built successfully after one time retry.
+        when(mockProvider.checkInitializedBefore(any()))
+                .thenReturn(CompletableFuture.completedFuture(true));
+
+        when(mockProvider.newPendingAckStore(any()))
+                // First, the method newPendingAckStore will fail with a retryable exception.
+                .thenReturn(FutureUtil.failedFuture(new ManagedLedgerException("mock fail new store")))
+                // Then, the method will be executed successfully.
+                .thenCallRealMethod();
+        transactionPendingAckStoreProviderField.set(pulsarServiceList.get(0), mockProvider);
+        @Cleanup
+        Consumer<byte[]> consumer3 = pulsarClient.newConsumer()
+                .subscriptionName("subName3")
+                .topic(topic)
+                .subscribe();
+
+        // 5. Test no-retryable exception:
+        // The consumer building will be failed without retrying.
```

Comment:
The consumer still retries because the error code from the broker side is `UnknownError`.

https://github.com/apache/pulsar/blob/6f9c8e7f70ec201d65c7fc270480bed3aa3b5aba/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ClientCnx.java#L794

and then the error code will be converted to a PulsarClientException, which is treated as retryable.

<img width="1558" alt="image" src="https://github.com/apache/pulsar/assets/18204803/2b09004d-0e2a-4e4e-9634-bb0c6971eebf">

Reply:
>The consumer still retries because the error code from the broker side is UnknownError.
Yes, it is correct. According to the discussion, manual intervention is necessary when an unexpected/unretryable exception happens. So  `UnknownError` may be acceptable.

Reply:
Another approach is to add new errors and exceptions to catch this case. What do you think about it?

Reply:
You can open another PR to add such new error (`TransactionComponentLoadFailedException`) as your comment in `exceptionHandleFuture` says. It's okay here for now. Your comment "retrying" here should only refer the server-side retry.

But since this test could take too long (48 seconds in my local env) to wait for the client side, it's better to reduce the timeout by creating `consumer4` with another `PulsarClient` that configures the operation timeout.

```java
        @Cleanup PulsarClient pulsarClient = PulsarClient.builder()
                .serviceUrl(pulsarServiceList.get(0).getBrokerServiceUrl())
                .operationTimeout(3, TimeUnit.SECONDS)
                .build();
```

Reply:
Good suggestion! Thanks